File: dev/sources/distro/Src-Utilities/rluawfx/source/genguid.cpp
/*********************************************************************
RLuaWfx
Copyright (C) 2004-2013 Roberto Rossi 
Web : http://www.redchar.net

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
*********************************************************************/

//#include "stdafx.h"
#include <Windows.h>
//#include <stdio.h>

/*
 * UnicodeToAnsi converts the Unicode string pszW to an ANSI string
 * and returns the ANSI string through ppszA. Space for the
 * the converted string is allocated by UnicodeToAnsi.
 */ 

HRESULT UnicodeToAnsi(LPCOLESTR pszW, char *buffer, int maxLen)
{

    ULONG cbAnsi, cCharacters;
    DWORD dwError;
	LPSTR ppszA;

    // If input is null then just return the same.
    if (pszW == NULL)
    {
        return NOERROR;
    }

    cCharacters = wcslen(pszW)+1;
    // Determine number of bytes to be allocated for ANSI string. An
    // ANSI string can have at most 2 bytes per character (for Double
    // Byte Character Strings.)
    cbAnsi = cCharacters*2;

    // Use of the OLE allocator is not required because the resultant
    // ANSI  string will never be passed to another COM component. You
    // can use your own allocator.
    ppszA = (LPSTR) CoTaskMemAlloc(cbAnsi);
    if (NULL == ppszA)
        return E_OUTOFMEMORY;

    // Convert to ANSI.
    if (0 == WideCharToMultiByte(CP_ACP, 0, pszW, cCharacters, ppszA,
                  cbAnsi, NULL, NULL))
    {
        dwError = GetLastError();
        CoTaskMemFree(ppszA);
        ppszA = NULL;
        return HRESULT_FROM_WIN32(dwError);
    }

	strncpy(buffer, ppszA,maxLen);
	CoTaskMemFree(ppszA);
    return NOERROR;

}
/*int _tmain(int argc, _TCHAR* argv[])
{

	GUID id;
	WCHAR buffer[255];
	char buffer_char[255];
	FILE *stream = NULL;
	int nguid;
	int i;

	if (argc > 1)
	{
		nguid = atoi(argv[1]);
		if (nguid > 0) {
			if (argc > 2)
				stream = fopen( argv[2], "w" );		

			//TODO : gestire msg errore se non si riesce ad aprire il file

			for (i=0;i<nguid;i++) {
				CoCreateGuid(&id);
				StringFromGUID2(id,buffer,254);
				UnicodeToAnsi(buffer,buffer_char,sizeof(buffer_char)-1);
				if (stream != NULL) {
					if (i>0)
						fprintf( stream, "\n", buffer_char);
					fprintf( stream, "%s", buffer_char);
				} else
					printf("%s\n", buffer_char); //se non esiste file, visualizza i valori				
			}

			if (stream != NULL) {
				fclose( stream );
			}//endif
		} else { //controllo nguid
			//TODO : msg errore per numero errato!
		}
	} else {
		//TODO : inserire help
	}
	return 0;
}*/

File: sources/distro/luascr/genguid.lua
--[[
Author  : Roberto Rossi
Version : 1.1.2
Web     : http://www.redchar.net

Questa procedura consente la generazione e l'inserimento di un certo numero
di GUID all'interno del file corrente

Copyright (C) 2012-2013 Roberto Rossi 
*******************************************************************************
This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
*******************************************************************************

Modelli generati :
1)
    // {87AF24BD-C9F9-4c69-B913-798429857EED}
    DEFINE_GUID(<<name>>, 
    0x87af24bd, 0xc9f9, 0x4c69, 0xb9, 0x13, 0x79, 0x84, 0x29, 0x85, 0x7e, 0xed);
2)
    // {87AF24BD-C9F9-4c69-B913-798429857EED}
    IMPLEMENT_OLECREATE(<<class>>, <<external_name>>, 
    0x87af24bd, 0xc9f9, 0x4c69, 0xb9, 0x13, 0x79, 0x84, 0x29, 0x85, 0x7e, 0xed);

3)
    // {87AF24BD-C9F9-4c69-B913-798429857EED}
    static const GUID <<name>> = 
    { 0x87af24bd, 0xc9f9, 0x4c69, { 0xb9, 0x13, 0x79, 0x84, 0x29, 0x85, 0x7e, 0xed } };

4)
    {87AF24BD-C9F9-4c69-B913-798429857EED}
    
]]


do
  require("luascr/rluawfx")
  
  --ritorna una delle sezioni che compongono la guid, in questo modo :
  -- parte 1 = primi 8 caratteri
  -- parte 2..3 = successivi 4 caratteri
  -- parte 4..11 = successivi 2 caratteri 
  -- esempio guid in ingresso = "{921C080F-0900-4AD1-ACCC-33F46C24EA23}"
  local function getPartGUID(guid, parte)
    local result = ""
    
    if (parte == 1) then
      result = string.sub(guid, 2, 9)
    elseif (parte==2) then
      result = string.sub(guid, 11, 14)
    elseif (parte==3) then
      result = string.sub(guid, 16, 19)
    elseif (parte==4) then
      result = string.sub(guid, 21, 22)
    elseif (parte==5) then
      result = string.sub(guid, 23, 24)
    elseif (parte==6) then
      result = string.sub(guid, 26, 27)
    elseif (parte==7) then
      result = string.sub(guid, 28, 29)
    elseif (parte==8) then
      result = string.sub(guid, 30, 31)
    elseif (parte==9) then
      result = string.sub(guid, 32, 33)
    elseif (parte==10) then
      result = string.sub(guid, 34, 35)
    elseif (parte==11) then
      result = string.sub(guid, 36, 37)
    end
    
    return "0x"..result
  end
  
  local function getDefineGUID(guid)
    --DEFINE_GUID(<<name>>, 0x87af24bd, 0xc9f9, 0x4c69, 0xb9, 0x13, 0x79, 0x84, 0x29, 0x85, 0x7e, 0xed);
    local result = "DEFINE_GUID(name, "
    result = result..getPartGUID(guid, 1)..", "
    result = result..getPartGUID(guid, 2)..", "
    result = result..getPartGUID(guid, 3)..", "
    result = result..getPartGUID(guid, 4)..", "
    result = result..getPartGUID(guid, 5)..", "
    result = result..getPartGUID(guid, 6)..", "
    result = result..getPartGUID(guid, 7)..", "
    result = result..getPartGUID(guid, 8)..", "
    result = result..getPartGUID(guid, 9)..", "
    result = result..getPartGUID(guid, 10)..", "
    result = result..getPartGUID(guid, 11)..");"
    
    return result
  end
  
  local function getOleGUID(guid)
    --IMPLEMENT_OLECREATE(<<class>>, <<external_name>>, 
    --0x87af24bd, 0xc9f9, 0x4c69, 0xb9, 0x13, 0x79, 0x84, 0x29, 0x85, 0x7e, 0xed);
    local result = "IMPLEMENT_OLECREATE(class, external_name, "
    result = result..getPartGUID(guid, 1)..", "
    result = result..getPartGUID(guid, 2)..", "
    result = result..getPartGUID(guid, 3)..", "
    result = result..getPartGUID(guid, 4)..", "
    result = result..getPartGUID(guid, 5)..", "
    result = result..getPartGUID(guid, 6)..", "
    result = result..getPartGUID(guid, 7)..", "
    result = result..getPartGUID(guid, 8)..", "
    result = result..getPartGUID(guid, 9)..", "
    result = result..getPartGUID(guid, 10)..", "
    result = result..getPartGUID(guid, 11)..");"
    
    return result
  end
  
  local function getConstGUID(guid)
    --static const GUID <<name>> = 
    --{ 0x87af24bd, 0xc9f9, 0x4c69, { 0xb9, 0x13, 0x79, 0x84, 0x29, 0x85, 0x7e, 0xed } };
    local result = "static const GUID name = { "
    result = result..getPartGUID(guid, 1)..", "
    result = result..getPartGUID(guid, 2)..", "
    result = result..getPartGUID(guid, 3)..", { "
    result = result..getPartGUID(guid, 4)..", "
    result = result..getPartGUID(guid, 5)..", "
    result = result..getPartGUID(guid, 6)..", "
    result = result..getPartGUID(guid, 7)..", "
    result = result..getPartGUID(guid, 8)..", "
    result = result..getPartGUID(guid, 9)..", "
    result = result..getPartGUID(guid, 10)..", "
    result = result..getPartGUID(guid, 11).." } };"
    
    return result
  end  
  
  --button ok
  function buttonOk_click(control, change)
    local val = wcl_strip:getValue("VAL")
    editor:ReplaceSel(val)
  end
  
  --main function
  local function main()
    local nguid
    local opzioni
    local testo
    local dest
    local GUIDs = {}
    
    dest = rfx_FN()
    
    if (rfx_GetGUID(dest)) then
      testo = string.lower(rfx_GF())
      
      GUIDs = { testo, 
                getDefineGUID(testo),
                getOleGUID(testo),
                getConstGUID(testo)
                }
      
      wcl_strip:init()      
      wcl_strip:addButtonClose()
      
      wcl_strip:addLabel(nil, _t(165))
      wcl_strip:addCombo("VAL")
      wcl_strip:addButton("OK",_t(251),buttonOk_click, true)
      
      wcl_strip:show()
      
      wcl_strip:setList("VAL", GUIDs)
      wcl_strip:setValue("VAL", GUIDs[1])
    end
  end
  
  main()
  
end

